tools/nbs_tools/calibrate_cameras.py

```diff
@@ -0,0 +1,507 @@
+
+
+copyreg.pickle(cv2.KeyPoint().__class__, _pickle_keypoints)
+
+
+def register(command):
+    command.help = "Calculate camera intrincs using an nbs recordings of checkered board patterns."
+
+    # Command arguments
+    command.add_argument(
+        "files", metavar="files", nargs="+", help="The nbs files to extract the compressed images from"
+    )
+    command.add_argument(
+        "--config",
+        "-c",
+        dest="config_path",
+        required=True,
+        help="The directory containing the configuration files for the cameras",
+    )
+    command.add_argument(
+        "--no-intrinsics",
+        action="store_true",
+        help="use the intrinsics provided in the configuration file and do not attempt to optimise them",
+    )
+    command.add_argument(
+        "--no-extrinsics",
+        action="store_true",
+        help="use the intrinsics provided in the configuration file and do not attempt to optimise them",
+    )
+
+
+def packetise_stream(decoder):
+
+    bytes_read = 0
+    for packet in decoder:
+
+        # Check for compressed images
+        if packet.type == "message.output.CompressedImage":
+
+            # Work out how many bytes we have read to get to this message
+            read = decoder.bytes_read() - bytes_read
+            bytes_read = decoder.bytes_read()
+
+            # Get some useful info into a pickleable format
+            yield {
+                "bytes_read": read,
+                "camera_name": packet.msg.name,
+                "timestamp": (packet.msg.timestamp.seconds, packet.msg.timestamp.nanos),
+                "data": packet.msg.data,
+                "format": packet.msg.format,
+            }
+
+
+def process_frame(item):
+    data = decode_image(item["data"], item["format"])
+
+    if len(data) == 1:
+        img = data[0]["image"].numpy()
+        fmt = data[0]["fourcc"]
+    else:
+        img = [d for d in data if d["name"] == "_colour"][0]["image"].numpy()
+        fmt = data[0]["fourcc"]
+
+    # Debayer if we need to
+    if fmt == fourcc("BGGR"):
+        img = cv2.cvtColor(img, cv2.COLOR_BayerBG2RGB)
+    elif fmt == fourcc("RGGB"):
+        img = cv2.cvtColor(img, cv2.COLOR_BayerRG2RGB)
+    elif fmt == fourcc("GRBG"):
+        img = cv2.cvtColor(img, cv2.COLOR_BayerGR2RGB)
+    elif fmt == fourcc("GBRG"):
+        img = cv2.cvtColor(img, cv2.COLOR_BayerGB2RGB)
+
+    params = cv2.SimpleBlobDetector_Params()
+
+    params.minThreshold = 10
+    params.maxThreshold = 200
+    params.thresholdStep = 10
+
+    params.minDistBetweenBlobs = 5
+
+    params.filterByConvexity = True
+    params.minConvexity = 0.95
+
+    # Filter by area going from a circle radius of about 2 pixels to one of about 1/22th of the width
+    params.filterByArea = True
+    params.minArea = 15
+    params.maxArea = math.pi * (img.shape[0] * (1 / 22)) ** 2
+
+    params.filterByCircularity = False
+
+    params.filterByInertia = False
```

Comment:
They change how the blob detector that is detecting the circles work, this could probably use some commenting. Several of those parameters I don't really understand myself.